# **Решение задач по библиотеке Pandas**

In [155]:
import pandas as pd

# **1. Подготовка исходных данных**

Считать файл 'transactions.csv' и записать его в таблицу transactions.

Таблица содержит следующие признаки:
- product_id - идентификатор продукта;
- customer_id	- идентификатор клиента;
- store_id - идентификатор магазина;
- month_of_year	- месяц покупки;
- store_sales - стоимость покупки.

In [156]:
transactions = pd.read_csv("transactions.csv") # Загрузка данных из файла transactions.csv в DataFrame

Считать файл 'products.csv' и записать его в таблицу products.

Таблица содержит следующие признаки:
- product_class_id - идентификатор категории продукта;
- product_id - идентификатор продукта;
- brand_name - производитель продукта;
- product_name - название продукта.

In [157]:
products = pd.read_csv("products.csv") # Загрузка данных из файла products.csv в DataFrame

Считать файл 'product_class.csv' и записать его в таблицу product_class.

Таблица содержит следующие признаки:
- product_class_id - идентификатор категории продукта;
- product_subcategory - название подкатегории продукта;
- product_category - название категории продукта;
- product_department - отдел магазина;
- product_family - продуктовые или продовольственные продукты.

In [158]:
product_class = pd.read_csv("product_class.csv") # Загрузка данных из файла product_class.csv в DataFrame prcl

# **2. Объединение таблиц**

## Метод concat()

На основе таблицы 'transactions' создать три таблицы: trans1 содержит первые 5000 строк, trans2 - строки с 5000 по 150000, trans3 - с 150000 и до конца таблицы.

In [159]:
trans1 = transactions.loc[:4999] # trans1 содержит первые 5000 записей (индексы от 0 до 4999)

In [160]:
trans2 = transactions.loc[5000:149999] # trans2 содержит записи с индексами от 5000 до 149999

In [161]:
trans3 = transactions.loc[150000:] # trans3 содержит записи начиная с индекса 150000 до конца DataFrame

Написать код, объединяющий таблицы 'trans1' и 'trans2' с использованием метода `concat()`.

In [162]:
conc = pd.concat([trans1, trans2]) # Склейка DataFrame trans1 и trans2 в один DataFrame conc
conc # Вывод DataFrame conc

product_id  customer_id  store_id  month_of_year  store_sales
0                1          157        24             12         8.55
1                1          456        15              6        11.40
2                1          638        11              9         8.55
3                1          916         7              4        11.40
4                1          923        15              7         8.55
...            ...          ...       ...            ...          ...
149995        1136            0      3239              3        14.44
149996        1136            0      3271              2        14.44
149997        1136            0      3499              1        10.83
149998        1136         1223      4391              6        10.83
149999        1136            0      4498             11        14.44

[150000 rows x 5 columns]

Написать код, объединяющий таблицы 'trans1', 'trans2' и 'trans3' с использованием метода `concat()`.

In [163]:
trans123 = pd.concat([trans1, trans2, trans3]) # Склейка DataFrame-ов trans1, trans2 и trans3 в один DataFrame trans123
trans123 # Вывод DataFrame trans123

product_id  customer_id  store_id  month_of_year  store_sales
0                1          157        24             12         8.55
1                1          456        15              6        11.40
2                1          638        11              9         8.55
3                1          916         7              4        11.40
4                1          923        15              7         8.55
...            ...          ...       ...            ...          ...
173597        1559            0      8653              2         3.24
173598        1559            0      8659             12         6.48
173599        1559            0      9271              5         6.48
173600        1559            0      9318              9         4.86
173601        1559            0      9851              2         4.86

[173602 rows x 5 columns]

Написать код, добавляющий в таблицу 'transactions' название продукта с использованием метода `concat()`. Исходную таблицу 'transactions' **перезаписывать не нужно**.

Написать вывод о возможности применения метода `concat()` для решения задачи и объяснить свое решение.

In [164]:
pd.concat([transactions, products["product_name"]], axis=1)

product_id  customer_id  store_id  month_of_year  store_sales  \
0                1          157        24             12         8.55   
1                1          456        15              6        11.40   
2                1          638        11              9         8.55   
3                1          916         7              4        11.40   
4                1          923        15              7         8.55   
...            ...          ...       ...            ...          ...   
173597        1559            0      8653              2         3.24   
173598        1559            0      8659             12         6.48   
173599        1559            0      9271              5         6.48   
173600        1559            0      9318              9         4.86   
173601        1559            0      9851              2         4.86   

                       product_name  
0            Washington Berry Juice  
1            Washington Mango Drink  
2       Washington Strawberry Drink  
3             Washington Cream Soda  
4              Washington Diet Soda  
...                             ...  
173597                          NaN  
173598                          NaN  
173599                          NaN  
173600                          NaN  
173601                          NaN  

[173602 rows x 6 columns]

ВЫВОД: Задача не решена. concat добавляет определенное количество строк. Когда строки в добавляемой таблице заканчиваются, оставляет занчения NaN. Таким образом значения в  добавленном стобце не зависят от product_id

## Метод merge()

Написать код, добавляющий в таблицу 'transactions' название продукта с использованием метода `merge()`. Результат записать в таблицу 'm1'.

Написать вывод о возможности применения метода `merge()` для решения задачи и объяснить свое решение.

Какой тип объединения необходимо использовать для решения задачи? Объяснить сделанный выбор.

In [165]:
cols = products[['product_id', 'product_name']] # Выбор только столбцов 'product_id' и 'product_name' из DataFrame
m1 = transactions.merge(cols, on='product_id', how='left') # Объединение DataFrame tr с DataFrame cols по столбцу 'product_id' с использованием левого соединения
m1 # Вывод строк объединенного DataFrame


product_id  customer_id  store_id  month_of_year  store_sales  \
0                1          157        24             12         8.55   
1                1          456        15              6        11.40   
2                1          638        11              9         8.55   
3                1          916         7              4        11.40   
4                1          923        15              7         8.55   
...            ...          ...       ...            ...          ...   
173597        1559            0      8653              2         3.24   
173598        1559            0      8659             12         6.48   
173599        1559            0      9271              5         6.48   
173600        1559            0      9318              9         4.86   
173601        1559            0      9851              2         4.86   

                  product_name  
0       Washington Berry Juice  
1       Washington Berry Juice  
2       Washington Berry Juice  
3       Washington Berry Juice  
4       Washington Berry Juice  
...                        ...  
173597     CDR Apple Preserves  
173598     CDR Apple Preserves  
173599     CDR Apple Preserves  
173600     CDR Apple Preserves  
173601     CDR Apple Preserves  

[173602 rows x 6 columns]

ВЫВОД: Задача решена. merge добавляет строки основываясь на объединении по одинаковым значениям в таблицах

Вывести размер и первые 2 строки таблиц 'products' и 'product_class'.

In [166]:
print(products.shape, product_class.shape) # вывод размерности (количество строк и столбцов) DataFrame pr и DataFrame prcl

(1560, 4) (109, 5)


In [167]:
products.head(2) # Вывод первых 2 строк

product_class_id  product_id  brand_name            product_name
0                30           1  Washington  Washington Berry Juice
1                52           2  Washington  Washington Mango Drink

In [168]:
product_class.head(2) # Вывод первых 2 строк

product_class_id product_subcategory product_category product_department  \
0                 1                Nuts        Specialty            Produce   
1                 2           Shellfish          Seafood            Seafood   

  product_family  
0           Food  
1           Food

Написать код, добавляющий в таблицу 'products' данные о подкатегории ('product_subcategory') и категории ('product_category') для каждого продукта с использованием метода `marge()` с агрументом `how = 'inner'`. Результат записать в таблицу 'm2_inner'.


In [169]:
pspc = product_class[['product_class_id', 'product_subcategory', 'product_category']] # Создание DataFrame pspc путем выбора столбцов 'product_class_id', 'product_subcategory', и 'product_category' из DataFrame prcl

m2_inner = products.merge(pspc, on='product_class_id', how='inner') # Объединение DataFrame pr с DataFrame pspc по столбцу 'product_class_id' с использованием внутреннего соединения
m2_inner # Вывод DataFrame m2_inner

product_class_id  product_id  brand_name  \
0                   30           1  Washington   
1                   52           2  Washington   
2                   52           3  Washington   
3                   19           4  Washington   
4                   19           5  Washington   
...                ...         ...         ...   
1550                53        1556         CDR   
1551                33        1557         CDR   
1552                53        1558         CDR   
1553                33        1559         CDR   
1554                31        1560         CDR   

                        product_name product_subcategory      product_category  
0             Washington Berry Juice               Juice  Pure Juice Beverages  
1             Washington Mango Drink     Flavored Drinks                Drinks  
2        Washington Strawberry Drink     Flavored Drinks                Drinks  
3              Washington Cream Soda                Soda  Carbonated Beverages  
4               Washington Diet Soda                Soda  Carbonated Beverages  
...                              ...                 ...                   ...  
1550        CDR Creamy Peanut Butter       Peanut Butter      Jams and Jellies  
1551        CDR Strawberry Preserves           Preserves      Jams and Jellies  
1552  CDR Extra Chunky Peanut Butter       Peanut Butter      Jams and Jellies  
1553             CDR Apple Preserves           Preserves      Jams and Jellies  
1554                 CDR Grape Jelly               Jelly      Jams and Jellies  

[1555 rows x 6 columns]

Найти количество пропущенных значений для каждого признака в результирующей таблице 'm2'.

In [170]:
m2_inner.isna().sum() # Вычисление количества пропущенных значений в каждом столбце DataFrame

product_class_id       0
product_id             0
brand_name             0
product_name           0
product_subcategory    0
product_category       0
dtype: int64

Почему в результирующей таблице 'm2_inner' количество записей уменьшилось с 1560 до 1555?

ВЫВОД: Потому что inner добавляет в результирующую таблицу только те строки, в которых нашлось соответсвие в двух таблицах одновременно

Добавить в таблицу 'products' данные о подкатегории ('product_subcategory') и категории ('product_category') для каждого продукта с использованием метода `marge()` с агрументом `how = 'left'`. Записать полученный результат в таблицу 'm2_left'.


In [171]:
m2_left = products.merge(pspc, on='product_class_id', how='left') # Объединение DataFrame pr с DataFrame pspc по столбцу 'product_class_id' с использованием левого соединения
m2_left # Вывод DataFrame

product_class_id  product_id  brand_name  \
0                   30           1  Washington   
1                   52           2  Washington   
2                   52           3  Washington   
3                   19           4  Washington   
4                   19           5  Washington   
...                ...         ...         ...   
1555                53        1556         CDR   
1556                33        1557         CDR   
1557                53        1558         CDR   
1558                33        1559         CDR   
1559                31        1560         CDR   

                        product_name product_subcategory      product_category  
0             Washington Berry Juice               Juice  Pure Juice Beverages  
1             Washington Mango Drink     Flavored Drinks                Drinks  
2        Washington Strawberry Drink     Flavored Drinks                Drinks  
3              Washington Cream Soda                Soda  Carbonated Beverages  
4               Washington Diet Soda                Soda  Carbonated Beverages  
...                              ...                 ...                   ...  
1555        CDR Creamy Peanut Butter       Peanut Butter      Jams and Jellies  
1556        CDR Strawberry Preserves           Preserves      Jams and Jellies  
1557  CDR Extra Chunky Peanut Butter       Peanut Butter      Jams and Jellies  
1558             CDR Apple Preserves           Preserves      Jams and Jellies  
1559                 CDR Grape Jelly               Jelly      Jams and Jellies  

[1560 rows x 6 columns]

Найти количество пропущенных значений для каждого признака результирующей таблицы 'm2_left'.

In [172]:
m2_left.isna().sum() # Вычисление количества пропущенных значений в каждом столбце DataFrame

product_class_id       0
product_id             0
brand_name             0
product_name           0
product_subcategory    5
product_category       5
dtype: int64

Вывести записи таблицы 'm2_left', в которых есть пропущенные значения.

In [173]:
m2_left[m2_left.isna().any(axis=1)] # Фильтрация DataFrame m2_left для отображения только строк, содержащих хотя бы одно пропущенное значение

product_class_id  product_id    brand_name  \
158                 86         159         Denny   
469                 86         470      Red Wing   
780                 86         781     Cormorant   
1096                86        1097  High Quality   
1407                86        1408        Sunset   

                          product_name product_subcategory product_category  
158          Denny Copper Pot Scrubber                 NaN              NaN  
469       Red Wing Copper Pot Scrubber                 NaN              NaN  
780      Cormorant Copper Pot Scrubber                 NaN              NaN  
1096  High Quality Copper Pot Scrubber                 NaN              NaN  
1407        Sunset Copper Pot Scrubber                 NaN              NaN

Сделать вывод о причинах пропусков в результирующей таблице 'm2_left'.

ВЫВОД: Если при left в правой таблице не находится соответствующего значения для левой таблицы, в результирующую таблицу записывается NaN

Добавить в таблицу 'products' данные о подкатегории ('product_subcategory') и категории ('product_category') для каждого продукта с использованием метода `marge()` с агрументом `how = 'right'`. Записать полученный результат в таблицу 'm2_right'.


In [174]:
m2_right = products.merge(pspc, on='product_class_id', how='right') # Объединение DataFrame pr с DataFrame pspc по столбцу 'product_class_id' с использованием правого соединения
m2_right # Вывод DataFrame m2_right

product_class_id  product_id  brand_name                 product_name  \
0                    1       195.0    High Top             High Top Almonds   
1                    1       205.0    High Top          High Top Mixed Nuts   
2                    1       206.0    High Top      High Top Canned Peanuts   
3                    1       211.0    High Top             High Top Walnuts   
4                    1       214.0    High Top          High Top Party Nuts   
...                ...         ...         ...                          ...   
1558               110      1050.0     PigTail      PigTail Lemon Popsicles   
1559               110      1332.0  Carrington  Carrington Orange Popsicles   
1560               110      1349.0  Carrington   Carrington Grape Popsicles   
1561               110      1351.0  Carrington    Carrington Lime Popsicles   
1562               110      1361.0  Carrington   Carrington Lemon Popsicles   

     product_subcategory product_category  
0                   Nuts        Specialty  
1                   Nuts        Specialty  
2                   Nuts        Specialty  
3                   Nuts        Specialty  
4                   Nuts        Specialty  
...                  ...              ...  
1558           Popsicles  Frozen Desserts  
1559           Popsicles  Frozen Desserts  
1560           Popsicles  Frozen Desserts  
1561           Popsicles  Frozen Desserts  
1562           Popsicles  Frozen Desserts  

[1563 rows x 6 columns]

Найти количество пропущенных значений для каждого признака результирующей таблицы 'm2_right'.

In [175]:
m2_right.isna().sum() # Вычисление количества пропущенных значений в каждом столбце DataFrame

product_class_id       0
product_id             8
brand_name             8
product_name           8
product_subcategory    0
product_category       0
dtype: int64

Вывести записи результирующей таблицы 'm2_right', в которых есть пропущенные значения.

In [176]:
m2_right[m2_right.isna().any(axis=1)] # Фильтрация DataFrame m2_right для отображения только строк, содержащих хотя бы одно пропущенное значение

product_class_id  product_id brand_name product_name  \
83                   7         NaN        NaN          NaN   
204                 13         NaN        NaN          NaN   
300                 21         NaN        NaN          NaN   
316                 23         NaN        NaN          NaN   
829                 59         NaN        NaN          NaN   
1080                68         NaN        NaN          NaN   
1081                69         NaN        NaN          NaN   
1322                92         NaN        NaN          NaN   

     product_subcategory   product_category  
83                Coffee          Dry Goods  
204     Fresh Vegetables         Vegetables  
300             Cleaners  Cleaning Supplies  
316          Shower Soap  Cleaning Supplies  
829      Dehydrated Soup      Packaged Soup  
1080           Dish Soap  Bathroom Products  
1081     Dishwasher Soap  Bathroom Products  
1322  Non-Alcoholic Wine      Beer and Wine

Сделать вывод о причинах пропущенных значений.

ВЫВОД: В правой таблице не нашлось соответсвий из левой таблицы, из-за чего при right соединении значения из левой таблицы были пропущены

Добавить в таблицу 'products' данные о подкатегории ('product_subcategory') и категории ('product_category') для каждого продукта с использованием метода `marge()` с агрументом `how = 'outer'`. Записать полученный результат в таблицу 'm2_outer'.


In [177]:
m2_outer = products.merge(pspc, on='product_class_id', how='outer') # Объединение DataFrame pr с DataFrame pspc по столбцу 'product_class_id' с использованием внешнего соединения (outer join)
m2_outer # Вывод DataFrame m2_outer

product_class_id  product_id  brand_name                 product_name  \
0                    1       195.0    High Top             High Top Almonds   
1                    1       205.0    High Top          High Top Mixed Nuts   
2                    1       206.0    High Top      High Top Canned Peanuts   
3                    1       211.0    High Top             High Top Walnuts   
4                    1       214.0    High Top          High Top Party Nuts   
...                ...         ...         ...                          ...   
1563               110      1050.0     PigTail      PigTail Lemon Popsicles   
1564               110      1332.0  Carrington  Carrington Orange Popsicles   
1565               110      1349.0  Carrington   Carrington Grape Popsicles   
1566               110      1351.0  Carrington    Carrington Lime Popsicles   
1567               110      1361.0  Carrington   Carrington Lemon Popsicles   

     product_subcategory product_category  
0                   Nuts        Specialty  
1                   Nuts        Specialty  
2                   Nuts        Specialty  
3                   Nuts        Specialty  
4                   Nuts        Specialty  
...                  ...              ...  
1563           Popsicles  Frozen Desserts  
1564           Popsicles  Frozen Desserts  
1565           Popsicles  Frozen Desserts  
1566           Popsicles  Frozen Desserts  
1567           Popsicles  Frozen Desserts  

[1568 rows x 6 columns]

Найти количество пропущенных значений для каждого признака результирующей таблицы 'm2_outer'.

In [178]:
m2_outer.isna().sum() # Вычисление количества пропущенных значений в каждом столбце DataFrame

product_class_id       0
product_id             8
brand_name             8
product_name           8
product_subcategory    5
product_category       5
dtype: int64

Вывести записи результирующей таблицы 'm2_outer', в которых есть пропущенные значения.

In [179]:
m2_right[m2_right.isna().any(axis=1)] # Фильтрация DataFrame m2_outer для отображения только строк, содержащих хотя бы одно пропущенное значение

product_class_id  product_id brand_name product_name  \
83                   7         NaN        NaN          NaN   
204                 13         NaN        NaN          NaN   
300                 21         NaN        NaN          NaN   
316                 23         NaN        NaN          NaN   
829                 59         NaN        NaN          NaN   
1080                68         NaN        NaN          NaN   
1081                69         NaN        NaN          NaN   
1322                92         NaN        NaN          NaN   

     product_subcategory   product_category  
83                Coffee          Dry Goods  
204     Fresh Vegetables         Vegetables  
300             Cleaners  Cleaning Supplies  
316          Shower Soap  Cleaning Supplies  
829      Dehydrated Soup      Packaged Soup  
1080           Dish Soap  Bathroom Products  
1081     Dishwasher Soap  Bathroom Products  
1322  Non-Alcoholic Wine      Beer and Wine

Сделать вывод о причинах пропущенных значений.

Сделать общий вывод о том, какой (какие) тип соединения можно использовать для решения поставленной задачи.

ВЫВОД: Все значения левой таблицы не имеющие соответвия в правой таблице и все значения правой таблицы не имеющие соответсвия в левой таблице обрели значения NaN. Можно сравнить с left + right соединением

Написать код, добаляющий в таблицу 'transactions' категорию каждого проданного продукта.

Исходную таблицу 'transactions' **перезаписывать не нужно**.

In [180]:
tmp = transactions.merge(products[['product_id', 'product_class_id']].merge(product_class[['product_class_id', 'product_category']], on='product_class_id', how='inner')[['product_id', 'product_category']], on='product_id', how='inner')
tmp
# Объединение DataFrame pr с DataFrame, содержащим только столбцы 'product_id' и 'product_class_id' из DataFrame products,
# затем результат этого объединения объединяется с DataFrame, содержащим только столбцы 'product_class_id' и 'product_category' из DataFrame products_class,
# и фильтрация оставляет только столбцы 'product_id' и 'product_category'
# Вывод результата

product_id  customer_id  store_id  month_of_year  store_sales  \
0                1          157        24             12         8.55   
1                1          456        15              6        11.40   
2                1          638        11              9         8.55   
3                1          916         7              4        11.40   
4                1          923        15              7         8.55   
...            ...          ...       ...            ...          ...   
173125        1559            0      8653              2         3.24   
173126        1559            0      8659             12         6.48   
173127        1559            0      9271              5         6.48   
173128        1559            0      9318              9         4.86   
173129        1559            0      9851              2         4.86   

            product_category  
0       Pure Juice Beverages  
1       Pure Juice Beverages  
2       Pure Juice Beverages  
3       Pure Juice Beverages  
4       Pure Juice Beverages  
...                      ...  
173125      Jams and Jellies  
173126      Jams and Jellies  
173127      Jams and Jellies  
173128      Jams and Jellies  
173129      Jams and Jellies  

[173130 rows x 6 columns]

# **3. Агрегирование и группировка данных**

Объеденить таблицы 'transactions' и 'products' с использованием метода `merge()` и записать результат в таблицу df.

In [181]:
df = transactions.merge(products) # Объединение DataFrame transactions с DataFrame products
df # Вывод результата

product_id  customer_id  store_id  month_of_year  store_sales  \
0                1          157        24             12         8.55   
1                1          456        15              6        11.40   
2                1          638        11              9         8.55   
3                1          916         7              4        11.40   
4                1          923        15              7         8.55   
...            ...          ...       ...            ...          ...   
173597        1559            0      8653              2         3.24   
173598        1559            0      8659             12         6.48   
173599        1559            0      9271              5         6.48   
173600        1559            0      9318              9         4.86   
173601        1559            0      9851              2         4.86   

        product_class_id  brand_name            product_name  
0                     30  Washington  Washington Berry Juice  
1                     30  Washington  Washington Berry Juice  
2                     30  Washington  Washington Berry Juice  
3                     30  Washington  Washington Berry Juice  
4                     30  Washington  Washington Berry Juice  
...                  ...         ...                     ...  
173597                33         CDR     CDR Apple Preserves  
173598                33         CDR     CDR Apple Preserves  
173599                33         CDR     CDR Apple Preserves  
173600                33         CDR     CDR Apple Preserves  
173601                33         CDR     CDR Apple Preserves  

[173602 rows x 8 columns]

Определить количество уникальных категорий продуктов ('product_class_id').

In [182]:
products["product_class_id"].unique() # Вывод уникальных значений из столбца "product_class_id" в DataFrame

array([ 30,  52,  19,  35,  62,  58,  93,  94,  95,  96,  97,  98,   3,
        89, 103, 102,  15,  14,  11,  76,   6,  81,  91,  16,  77,   8,
        66,  34, 110, 100,  29,  63,  64,  60,  10,  28,   9,  78,  39,
        79,  88,  70,  71,  72,  73,  75,  74,  80,  56, 101,  55,  22,
        18,  20,  40,  87,  47,  86,  85,  41,  42,  65, 104, 108, 107,
       105, 106,  61,  99,  67,   1,  24,   2,  45,  12,  82,  54,  83,
        46,  84,  17, 109,  36,  37,  38,   5,  57,  43,  44,  27,  26,
        25,  49,  53,  51,  50,  90,  48,  33,  31,   4,  32], dtype=int64)

Определить количество клиентов.

In [183]:
df['customer_id'].unique().shape # Вывод количества уникальных значений из столбца "customer_id"

(5653,)

Определить количество покупок, приходящихся на каждый месяц.

In [184]:
transactions[['month_of_year', 'store_sales']].groupby('month_of_year').count()
# Группировка данных в DataFrame tr по месяцу (столбец 'month_of_year')
# и подсчет количества продаж (записей в столбце 'store_sales') для каждого месяца

store_sales
month_of_year             
1                    14064
2                    13682
3                    15414
4                    13170
5                    13726
6                    13818
7                    15502
8                    14074
9                    13316
10                   12950
11                   16460
12                   17426

Определить долю покупок, приходящихся на каждый месяц.

In [185]:
s = transactions.shape[0] # Получаем количество строк в transactions
tmp = transactions[['month_of_year', 'store_sales']].groupby('month_of_year').count().reset_index() # Таблица из прошлого задания с обновленными индексами
tmp['part'] = tmp['store_sales'] / s # Столбец store_sales на кол-во строк
tmp # Вывод результата

month_of_year  store_sales      part
0               1        14064  0.081013
1               2        13682  0.078812
2               3        15414  0.088789
3               4        13170  0.075863
4               5        13726  0.079066
5               6        13818  0.079596
6               7        15502  0.089296
7               8        14074  0.081070
8               9        13316  0.076704
9              10        12950  0.074596
10             11        16460  0.094815
11             12        17426  0.100379

Разделить диапазон признака 'store_sales' на 5 интервалов и определить количество покупок в каждом интервале.

In [186]:
tmp = pd.cut(transactions['store_sales'], bins=5) # Разделяем столбец store_sales на 5 равных по размеру интервалов. Можно настроить количество и пределы интервалов, передавая список с конкретными границами.
tmp.value_counts().sort_index() # Подсчёт количества записей в каждом интервале и сортировка результатов по интервалам.

store_sales
(0.47, 6.512]       94320
(6.512, 12.524]     69332
(12.524, 18.536]     9590
(18.536, 24.548]      352
(24.548, 30.56]         8
Name: count, dtype: int64

Вычислить минимальную, максимальную и среднюю стоимость покупок (store_sales) с использованием метода `agg()`.

In [187]:
transactions[['store_sales']].agg(['min', 'max', 'mean']) # Вычисление минимального, максимального и среднего значений столбца 'store_sales' в DataFrame 

store_sales
min      0.500000
max     30.560000
mean     6.511619

Вычислить сумму, среднее значение и количество покупок для каждого клиента с использованием метода `agg()`.

In [188]:
tmp = transactions[['customer_id', 'store_sales']].groupby('customer_id').agg(['sum', 'count']) # Группировка данных в DataFrame transactions по 'customer_id' с агрегацией суммы и количества записей столбца 'store_sales'
tmp['sred'] = tmp['store_sales']['sum'] / tmp['store_sales']['count'] # Расчет средней суммы покупки каждого покупателя
tmp # Вывод итогового DataFrame

store_sales              sred
                    sum  count           
customer_id                              
0             414003.82  63627   6.506732
3                 98.54     17   5.796471
5                  1.08      1   1.080000
6                 65.76      6  10.960000
10                53.82      9   5.980000
...                 ...    ...        ...
10268             22.26      2  11.130000
10273            329.83     48   6.871458
10274            191.09     24   7.962083
10275             44.02      7   6.288571
10277             29.96      5   5.992000

[5653 rows x 3 columns]

Для каждого клиента вывести список названий продуктов всех его покупок и количество покупок  с использованием метода `agg()` и записать в таблицу 'purch'.

In [189]:
purch = df[['customer_id', 'product_name']].groupby('customer_id').agg([lambda x: list(set(x)), 'count']).reset_index() # Группировка данных по 'customer_id' и агрегация уникальных названий продуктов и их количества и сброс индекса для возвращения 'customer_id' в столбцы DataFrame
purch # Вывод результата

customer_id                                       product_name       
                                                         <lambda_0>  count
0              0  [Imagine Pancake Mix, Gauss Monthly Auto Magaz...  63627
1              3  [Pleasant Canned Tuna in Water, Monarch Spaghe...     17
2              5                               [Moms Sliced Turkey]      1
3              6  [BBB Best Strawberry Preserves, Modell Cranber...      6
4             10  [Club Whole Milk, Dollar Monthly Computer Maga...      9
...          ...                                                ...    ...
5648       10268      [Nationeel Potato Chips, Skinner Mango Drink]      2
5649       10273  [Carrington Frozen Pepperoni Pizza, Dollar Mon...     48
5650       10274  [Carrington Frozen Pepperoni Pizza, Carlson Mi...     24
5651       10275  [Ebony Asparagus, Colony Pumpernickel Bread, T...      7
5652       10277  [Best Choice No Salt Popcorn, Ebony Sweet Peas...      5

[5653 rows x 3 columns]

В таблице 'purch' переименовать столбцы на 'purchases' и 'num_purhcases'.

In [190]:
purch.columns = ['customer_id', 'purchases', 'num_purchases'] # Переименование столбцов
purch # Вывод результата

customer_id                                          purchases  \
0               0  [Imagine Pancake Mix, Gauss Monthly Auto Magaz...   
1               3  [Pleasant Canned Tuna in Water, Monarch Spaghe...   
2               5                               [Moms Sliced Turkey]   
3               6  [BBB Best Strawberry Preserves, Modell Cranber...   
4              10  [Club Whole Milk, Dollar Monthly Computer Maga...   
...           ...                                                ...   
5648        10268      [Nationeel Potato Chips, Skinner Mango Drink]   
5649        10273  [Carrington Frozen Pepperoni Pizza, Dollar Mon...   
5650        10274  [Carrington Frozen Pepperoni Pizza, Carlson Mi...   
5651        10275  [Ebony Asparagus, Colony Pumpernickel Bread, T...   
5652        10277  [Best Choice No Salt Popcorn, Ebony Sweet Peas...   

      num_purchases  
0             63627  
1                17  
2                 1  
3                 6  
4                 9  
...             ...  
5648              2  
5649             48  
5650             24  
5651              7  
5652              5  

[5653 rows x 3 columns]

Отсортировать записи в таблице 'purch' по возрастанию количества покупок.

In [191]:
purch.sort_values(by='num_purchases') # Сортировка строк

customer_id                                          purchases  \
4813         8689                                [Bravo Canned Peas]   
1886         3346                            [Sunset Silver Cleaner]   
1347         2374                       [Gulf Coast Tasty Candy Bar]   
2405         4293                               [High Top Cantelope]   
1349         2377                               [Atomic Spicy Mints]   
...           ...                                                ...   
453           829  [Red Wing Tissues, Imagine Pancake Mix, Consol...   
277           523  [CDR Creamy Peanut Butter, Nationeel Chocolate...   
536           980  [Red Wing Tissues, CDR Creamy Peanut Butter, C...   
261           504  [Moms Sliced Ham, CDR Creamy Peanut Butter, Pl...   
0               0  [Imagine Pancake Mix, Gauss Monthly Auto Magaz...   

      num_purchases  
4813              1  
1886              1  
1347              1  
2405              1  
1349              1  
...             ...  
453             457  
277             626  
536             731  
261             869  
0             63627  

[5653 rows x 3 columns]

Определить сумму покупок для каждого магазина ('store_id').

Признак 'store_id' преобразовать из индекса в столбец.

In [192]:
transactions.groupby('store_id')['store_sales'].sum().reset_index()
# Группируем данные по магазинам с помощью метода groupby.
# Суммируем значения столбца store_sales в каждой группе с помощью метода sum.
# Сбрасываем индекс таблицы с помощью метода reset_index.

store_id  store_sales
0            2      4739.23
1            3     52976.60
2            5         1.08
3            6     45816.00
4            7     54540.42
...        ...          ...
5584     10268        22.26
5585     10273       329.83
5586     10274       191.09
5587     10275        44.02
5588     10277        29.96

[5589 rows x 2 columns]

Определить среднюю стоимость покупок в различных магазинах ('store_id') для продуктов различных брендов ('brand_name').

In [193]:
transactions[['store_id', 'store_sales', 'product_id']].merge(products[['product_id', 'brand_name']], on='product_id', how='inner')[['store_id', 'store_sales', 'brand_name']].groupby('brand_name').agg(sales_mean=('store_sales', 'mean'), stores=('store_id', lambda x: list(set(x))))
# Выборка необходимых столбцов из транзакций и объединение с данными о брендах продуктов
# Группировка по бренду, расчёт средних продаж и получение списка уникальных магазинов
# Среднее значение продаж по бренду
# Список уникальных магазинов для каждого бренда
# Вывод результатов

sales_mean                                             stores
brand_name                                                               
ADJ           7.804138  [5376, 1153, 2, 3, 5123, 6, 7, 1159, 2441, 986...
Akron         5.341043  [2, 3, 4100, 6, 7, 11, 13, 15, 16, 17, 5650, 9...
American      6.801710  [4096, 10240, 2, 3, 6148, 2053, 6, 7, 4099, 61...
Amigo         9.016870  [2, 3, 6, 7, 9224, 4618, 11, 2572, 13, 14, 15,...
Applause      7.974138  [2, 3, 4099, 6, 7, 6153, 11, 6155, 13, 14, 15,...
...                ...                                                ...
Toucan        4.293821  [2, 3, 6149, 6, 7, 9224, 4618, 11, 1548, 13, 1...
Tri-State     6.596892  [8193, 2, 3, 8194, 6, 7, 8202, 11, 13, 14, 15,...
Urban         6.770602  [10240, 9728, 2, 3, 4099, 5633, 6, 7, 10243, 6...
Walrus        7.274639  [2, 3, 6, 7, 10250, 11, 13, 14, 15, 16, 17, 22...
Washington    6.771649  [2, 3, 2052, 6, 7, 8202, 11, 6155, 13, 14, 15,...

[111 rows x 2 columns]

Определить среднюю стоимость покупок в различных магазинах ('store_id') для продуктов различных брендов ('brand_name').

Использовать механизм сводных таблиц (метод `pivot_table()`).

Объяснить, что означают `NaN` в полученной результирующей таблице.

In [194]:
tmp = transactions[['store_id', 'store_sales', 'product_id']].merge(products[['product_id', 'brand_name']], on='product_id', how='inner')
# Объединяем DataFrame `transactions`, содержащий 'store_id', 'store_sales', 'product_id',
# с DataFrame `products`, содержащим 'product_id' и 'brand_name'.
# Соединение происходит по 'product_id', тип соединения - внутреннее (inner), чтобы сохранить только те строки,
# в которых существует соответствие в обоих DataFrame.

pivot_table = pd.pivot_table(tmp, values='store_sales', index='store_id', columns='brand_name', aggfunc='mean')
# `pivot_table` используется для агрегации средней стоимости продаж (`store_sales`) по каждому магазину (`store_id`)
# и по каждому бренду (`brand_name`), что помогает анализировать средние продажи брендов в разных магазинах.

pivot_table # Вывод сводной таблицы


brand_name   ADJ     Akron  American      Amigo  Applause    Atomic  \
store_id                                                              
2           4.14  2.340000  3.456471   2.790000  5.293333  4.564000   
3           8.28  5.384615  7.078493   8.317273  8.872308  6.411731   
5            NaN       NaN       NaN        NaN       NaN       NaN   
6           7.59  6.033333  7.141639   9.380000  8.016667  5.930625   
7           8.97  4.655000  7.387532  10.207143  7.863636  6.826800   
...          ...       ...       ...        ...       ...       ...   
10268        NaN       NaN       NaN        NaN       NaN       NaN   
10273        NaN       NaN       NaN        NaN       NaN       NaN   
10274        NaN       NaN       NaN        NaN       NaN       NaN   
10275        NaN       NaN  5.040000        NaN       NaN       NaN   
10277        NaN       NaN       NaN        NaN       NaN       NaN   

brand_name   BBB Best      Best  Best Choice    Better  ...  Thresher  \
store_id                                                ...             
2            2.910400  2.735000     3.796923  2.278421  ...  3.047500   
3            5.445460  6.849286     6.739605  5.980887  ...  7.650909   
5                 NaN       NaN          NaN       NaN  ...       NaN   
6            6.450552  8.460500     7.300294  5.522763  ...  7.507500   
7            5.861000  7.433125     7.267586  5.552804  ...  6.402632   
...               ...       ...          ...       ...  ...       ...   
10268             NaN       NaN          NaN       NaN  ...       NaN   
10273       10.896667       NaN    10.600000       NaN  ...       NaN   
10274             NaN       NaN    15.560000  1.140000  ...       NaN   
10275             NaN       NaN          NaN       NaN  ...       NaN   
10277             NaN       NaN     4.760000       NaN  ...       NaN   

brand_name    Tip Top     Token  Top Measure  Toretti    Toucan  Tri-State  \
store_id                                                                     
2                 NaN  2.543846     2.325000      NaN  2.100000   3.004444   
3            7.187143  5.081857     5.702500     4.84  4.669412   6.516036   
5                 NaN       NaN          NaN      NaN       NaN        NaN   
6           10.438889  4.636949     5.431333     3.63  4.505000   6.564514   
7            6.158571  4.072931     5.497857     3.63  4.156667   6.590045   
...               ...       ...          ...      ...       ...        ...   
10268             NaN       NaN          NaN      NaN       NaN        NaN   
10273             NaN       NaN     2.100000      NaN       NaN   6.340000   
10274             NaN       NaN          NaN      NaN       NaN        NaN   
10275             NaN       NaN          NaN      NaN       NaN   1.860000   
10277             NaN       NaN          NaN      NaN       NaN        NaN   

brand_name     Urban    Walrus  Washington  
store_id                                    
2           7.480000  3.375000    4.096250  
3           7.195455  7.834894    7.374468  
5                NaN       NaN         NaN  
6           7.242400  7.278293    8.051064  
7           6.238519  7.408571    7.084815  
...              ...       ...         ...  
10268            NaN       NaN         NaN  
10273            NaN       NaN         NaN  
10274       6.500000       NaN    4.380000  
10275            NaN       NaN         NaN  
10277            NaN       NaN         NaN  

[5589 rows x 111 columns]

ВЫВОД: Если в исходном DataFrame нет записей о покупках для конкретного бренда в конкретном магазине, в соответствующей ячейке таблицы будет стоять значение NaN.

Определить количество продаж в каждом месяце ('month_of_year') для каждого бренда ('brand_name').

Использовать кросс-таблицы.

Пояснить, что означают числа в ячейках полученной таблицы.

In [195]:
tmp = transactions[['store_id', 'store_sales', 'product_id', 'month_of_year']].merge(products[['product_id', 'brand_name']], on='product_id', how='inner') # Создаем таблицу tmp путем объединения таблиц transactions и products по столбцу product_id.

cross_table = pd.crosstab(index=[tmp['brand_name']], columns=tmp['month_of_year'], values=tmp['store_sales'], aggfunc='sum')
# Создаем сводную таблицу cross_table с помощью метода crosstab библиотеки pandas.
# В качестве индекса используем столбец brand_name, в качестве столбцов - month_of_year.
# В качестве значений используем столбец store_sales, в качестве функции агрегации - сумма.

cross_table # Вывод DataFrame


month_of_year       1        2        3        4        5        6        7   \
brand_name                                                                     
ADJ              82.80    33.12    66.24   115.92    22.08    77.28    77.28   
Akron            80.72   111.92   178.40    62.96    49.16    80.56   108.52   
American        868.68   696.58  1105.60   570.42  1131.98   951.20  1096.72   
Amigo           205.16   191.96   155.42   109.08    75.26   159.98   248.58   
Applause         88.76   166.32   185.04    83.16   125.20   128.88    75.72   
...                ...      ...      ...      ...      ...      ...      ...   
Toucan           78.20    50.20    72.12   111.00    85.96    53.20    80.44   
Tri-State      2530.88  2478.04  2477.26  2096.72  2512.06  2386.64  2560.84   
Urban           332.08   303.46   381.84   338.48   327.42   433.52   414.06   
Walrus          429.30   453.80   656.60   481.06   455.18   533.04   760.42   
Washington      743.12   579.78   418.12   681.20   704.12   652.24   609.34   

month_of_year       8        9        10       11       12  
brand_name                                                  
ADJ              44.16   143.52    99.36   115.92    27.60  
Akron           122.40    66.20    63.08   136.56   167.96  
American        791.94   683.90   778.52  1166.20  1054.60  
Amigo           110.78   130.38    97.24   280.22   309.82  
Applause        230.84   190.56   158.88   205.52   211.12  
...                ...      ...      ...      ...      ...  
Toucan          117.04   122.32    72.00    69.40   144.40  
Tri-State      2498.62  2504.70  2649.64  2893.52  3271.34  
Urban           262.30   271.38   346.80   364.96   272.52  
Walrus          510.50   497.20   513.58   703.52   654.82  
Washington      603.44   618.12   610.32   773.12   808.02  

[111 rows x 12 columns]

ОБЪЯСНЕНИЕ: Числа в ячейках полученной кросс-таблицы представляют собой количество продаж конкретного бренда в каждом месяце. Например, если в ячейке на пересечении строки с брендом 'Amigo' и столбца с месяцем '6' стоит число 159.98, это означает, что в январе было продано 159.98 единиц продукции бренда 'Amigo'.

Сформировать таблицу df1, в которой будут записи о транзакциях тех групп, в которых средняя стоимость покупок каждого покупателя составляет не менее 13.

Использовать метод `filter()`.

In [196]:
tmp = transactions[['customer_id', 'store_sales']].groupby('customer_id').mean().reset_index()
# Выборка определенных столбцов из DataFrame `transactions` и группировка по 'customer_id'.
# Затем для каждой группы рассчитывается среднее значение 'store_sales'.
# После выполнения группировки и агрегации выполняется сброс индекса для возврата DataFrame.

df1 = tmp[tmp['store_sales'] >= 13]
# Фильтрация сформированного DataFrame, чтобы оставить только те строки,
# где средние 'store_sales' больше или равны 13.

df1 # Вывод отфильтрованного DataFrame

customer_id  store_sales
103           200       13.080
1263         2212       13.345
2355         4207       15.880
3426         6129       13.160
3951         7154       13.680
5412         9800       13.040

Добавить в таблицу df столбец 'store_sales_group'
с категорией суммы покупки:
- `small` - сумма покупки < 10;
- `average` - сумма покупки >= 10 и < 13;
- `large` - сумма покупки >= 13.

Использовать функцию `applay()`.

In [197]:
def check(x): # Определение функции check для классификации значений в столбце store_sales
    if x < 10:
        return 'small'
    elif x >= 10 and x < 13:
        return 'average'
    return 'large'

df['store_sales_group'] = df['store_sales'].apply(check) # Применение функции check к столбцу store_sales с использованием метода apply()

df # Вывод обновленного DataFrame

product_id  customer_id  store_id  month_of_year  store_sales  \
0                1          157        24             12         8.55   
1                1          456        15              6        11.40   
2                1          638        11              9         8.55   
3                1          916         7              4        11.40   
4                1          923        15              7         8.55   
...            ...          ...       ...            ...          ...   
173597        1559            0      8653              2         3.24   
173598        1559            0      8659             12         6.48   
173599        1559            0      9271              5         6.48   
173600        1559            0      9318              9         4.86   
173601        1559            0      9851              2         4.86   

        product_class_id  brand_name            product_name store_sales_group  
0                     30  Washington  Washington Berry Juice             small  
1                     30  Washington  Washington Berry Juice           average  
2                     30  Washington  Washington Berry Juice             small  
3                     30  Washington  Washington Berry Juice           average  
4                     30  Washington  Washington Berry Juice             small  
...                  ...         ...                     ...               ...  
173597                33         CDR     CDR Apple Preserves             small  
173598                33         CDR     CDR Apple Preserves             small  
173599                33         CDR     CDR Apple Preserves             small  
173600                33         CDR     CDR Apple Preserves             small  
173601                33         CDR     CDR Apple Preserves             small  

[173602 rows x 9 columns]

# **4. Операции со строковыми данными**

Считать файл 'imdbratings.csv' и записать его в таблицу 'film'.

In [198]:
film = pd.read_csv('imdbratings.csv') # Читаем файл imdbratings.csv с помощью метода read_csv библиотеки pandas.
film.head() # Выводим первые 5 строк таблицы с помощью метода head.

star_rating                     title content_rating   genre  duration  \
0          9.3  The Shawshank Redemption              R   Crime       142   
1          9.2             The Godfather              R   Crime       175   
2          9.1    The Godfather: Part II              R   Crime       200   
3          9.0           The Dark Knight          PG-13  Action       152   
4          8.9              Pulp Fiction              R   Crime       154   

                                         actors_list  
0  [u'Tim Robbins', u'Morgan Freeman', u'Bob Gunt...  
1    [u'Marlon Brando', u'Al Pacino', u'James Caan']  
2  [u'Al Pacino', u'Robert De Niro', u'Robert Duv...  
3  [u'Christian Bale', u'Heath Ledger', u'Aaron E...  
4  [u'John Travolta', u'Uma Thurman', u'Samuel L....

Преобразовать названия признаков так, чтобы они начинались с заглавной буквы.

In [199]:
film.columns = film.columns.str.title() # Изменяем названия столбцов таблицы film с помощью метода str.title.
film.head() # Выводим первые 5 строк таблицы с помощью метода head.

Star_Rating                     Title Content_Rating   Genre  Duration  \
0          9.3  The Shawshank Redemption              R   Crime       142   
1          9.2             The Godfather              R   Crime       175   
2          9.1    The Godfather: Part II              R   Crime       200   
3          9.0           The Dark Knight          PG-13  Action       152   
4          8.9              Pulp Fiction              R   Crime       154   

                                         Actors_List  
0  [u'Tim Robbins', u'Morgan Freeman', u'Bob Gunt...  
1    [u'Marlon Brando', u'Al Pacino', u'James Caan']  
2  [u'Al Pacino', u'Robert De Niro', u'Robert Duv...  
3  [u'Christian Bale', u'Heath Ledger', u'Aaron E...  
4  [u'John Travolta', u'Uma Thurman', u'Samuel L....

Для возможности сортировки фильмов по названию записать артикль 'The' в конце названия через запятую, например:
- было 'The Shawshank Redemption';
- должно стать 'Shawshank Redemption, The'

In [200]:
film['Title'] = film['Title'].apply(lambda x: f'{x[4:] + ', The'}' if x.startswith('The ') else x) # Изменяем названия фильмов в столбце Title с помощью метода apply.
film.head() # Выводим первые 5 строк таблицы с помощью метода head.

Star_Rating                      Title Content_Rating   Genre  Duration  \
0          9.3  Shawshank Redemption, The              R   Crime       142   
1          9.2             Godfather, The              R   Crime       175   
2          9.1    Godfather: Part II, The              R   Crime       200   
3          9.0           Dark Knight, The          PG-13  Action       152   
4          8.9               Pulp Fiction              R   Crime       154   

                                         Actors_List  
0  [u'Tim Robbins', u'Morgan Freeman', u'Bob Gunt...  
1    [u'Marlon Brando', u'Al Pacino', u'James Caan']  
2  [u'Al Pacino', u'Robert De Niro', u'Robert Duv...  
3  [u'Christian Bale', u'Heath Ledger', u'Aaron E...  
4  [u'John Travolta', u'Uma Thurman', u'Samuel L....

Поменять название жанра 'Sci-Fi' на 'Science fiction'.

In [201]:
film['Genre'] = film['Genre'].apply(lambda x: 'Science fiction' if x == 'Sci-Fi' else x) # Изменяем названия жанров в столбце Genre с помощью метода apply
film['Genre'].unique() # Выводим уникальные названия жанров с помощью метода unique.

array(['Crime', 'Action', 'Drama', 'Western', 'Adventure', 'Biography',
       'Comedy', 'Animation', 'Mystery', 'Horror', 'Film-Noir',
       'Science fiction', 'History', 'Thriller', 'Family', 'Fantasy'],
      dtype=object)

Вывести информацию о фильмах, в которых снимался Brad Pitt.

In [202]:
film[film['Actors_List'].str.contains('Brad Pitt')] # Ищем строки в столбце Actors_List, в которых содержится имя Brad Pitt с помощью метода str.contains.

Star_Rating                                              Title  \
9            8.9                                         Fight Club   
24           8.7                                              Se7en   
106          8.3                                            Snatch.   
114          8.3                               Inglourious Basterds   
264          8.1                                     Twelve Monkeys   
508          7.8               Curious Case of Benjamin Button, The   
577          7.8                                     Ocean's Eleven   
683          7.7                                               Fury   
776          7.6                                          Moneyball   
779          7.6  Interview with the Vampire: The Vampire Chroni...   
807          7.6  Assassination of Jesse James by the Coward Rob...   
826          7.5                                           Sleepers   
877          7.5                                Legends of the Fall   
901          7.5                                              Babel   

    Content_Rating      Genre  Duration  \
9                R      Drama       139   
24               R      Drama       127   
106              R     Comedy       102   
114              R  Adventure       153   
264              R    Mystery       129   
508          PG-13      Drama       166   
577          PG-13      Crime       116   
683              R     Action       134   
776          PG-13  Biography       133   
779              R     Horror       123   
807              R  Biography       160   
826              R      Crime       147   
877              R      Drama       133   
901              R      Drama       143   

                                           Actors_List  
9    [u'Brad Pitt', u'Edward Norton', u'Helena Bonh...  
24   [u'Morgan Freeman', u'Brad Pitt', u'Kevin Spac...  
106  [u'Jason Statham', u'Brad Pitt', u'Benicio Del...  
114       [u'Brad Pitt', u'Diane Kruger', u'Eli Roth']  
264  [u'Bruce Willis', u'Madeleine Stowe', u'Brad P...  
508  [u'Brad Pitt', u'Cate Blanchett', u'Tilda Swin...  
577  [u'George Clooney', u'Brad Pitt', u'Julia Robe...  
683   [u'Brad Pitt', u'Shia LaBeouf', u'Logan Lerman']  
776     [u'Brad Pitt', u'Robin Wright', u'Jonah Hill']  
779  [u'Brad Pitt', u'Tom Cruise', u'Antonio Bander...  
807   [u'Brad Pitt', u'Casey Affleck', u'Sam Shepard']  
826  [u'Robert De Niro', u'Kevin Bacon', u'Brad Pitt']  
877  [u'Brad Pitt', u'Anthony Hopkins', u'Aidan Qui...  
901  [u'Brad Pitt', u'Cate Blanchett', u'Gael Garc\...

# **5. Операции с датой и временем**

Создать объект класса дата/время `Timestamp` вида 16 апреля 2024 года 10 часов 20 минут на основе:
- перечня целых чисел;
- строковой записи даты;
- целого числа, представляющего Unix-время.

In [203]:
timestamp1 = pd.Timestamp(year=2024, month=4, day=16, hour=10, minute=20) # Создаем объект timestamp1
timestamp2 = pd.Timestamp('2024-04-16 10:20') # Создаем объект timestamp2
timestamp3 = pd.Timestamp(1713252000.0, unit='s', tz='UTC') # Создаем объект timestamp3

Создать объект дата/время `Timestamp` вида 16 апреля 2024 года 10 часов 20 минут с использованием метода `to_datetime()` на основе:
- перечня целых чисел;
- строковой записи даты;
- целого числа, представляющего Unix-время.

In [204]:
timestamp4 = pd.to_datetime([2024, 4, 16, 10, 20]) # Создаем объект timestamp4
timestamp5 = pd.to_datetime('2024-04-16 10:20') # Создаем объект timestamp5
timestamp6 = pd.to_datetime(1713252000, unit='s') # Создаем объект timestamp6

Создать объект класса `DatetimeIndex`,  включающий даты начиная с 16 апреля 2024 года по 19 апреля 2024 года включительно с периодичностью 1 день.

Использовать:
- метод `DatetimeIndex()`;
- метод `to_datetime()`;
- метод `date_range()`.

In [205]:
dates1 = pd.DatetimeIndex(['2024-04-16', '2024-04-17', '2024-04-18', '2024-04-19']) # Создаем объект dates1 
dates2 = pd.DatetimeIndex(pd.to_datetime(['2024-04-16', '2024-04-17', '2024-04-18', '2024-04-19'])) # Создаем объект dates2
dates3 = pd.date_range(start='2024-04-16', end='2024-04-19', freq='D') # Создаем объект dates3

Создать объект `Timedelta` с временным интервалом 1 день.

In [206]:
delta = pd.Timedelta(days=1) # Создаем объект delta типа Timedelta с помощью конструктора Timedelta.

Вывести с помощью созданного объекта дату следующего дня.

In [207]:
from datetime import datetime # Импортируем модуль datetime.
current = datetime.now() # Получаем текущую дату и время с помощью метода now.
next = current + delta # Получаем дату и время, наступающее через 1 день с помощью оператора +.
next # Выводим DataFrame

datetime.datetime(2024, 4, 18, 0, 28, 7, 173471)

Создать объект `TimedeltaIndex`, содержащий 15 временных дельт с интервалом 1 час.

Использовать:
- метод `TimedeltaIndex()`;
- метод `to_timedelta()`;
- метод `timedelta_range()`.

In [208]:
timedelta1 = pd.TimedeltaIndex([pd.Timedelta(hours=i) for i in range(15)]) # Создаем объект timedelta1
timedelta2 = pd.to_timedelta(['{} hours'.format(i) for i in range(15)]) # Создаем объект timedelta2
timedelta3 = pd.timedelta_range(start='0 hours', periods=15, freq='h') # Создаем объект timedelta3

Для выполнения следующих заданий используется набор данных о динамике велосипедного движения между центром города и некоторым его районом:
- `Time` - время измерения;
- `direction_centre`	- количество велосипедистов в направлении центра;
- `direction_region` - количество велосипедистов в направлении района.

Считать файл 'time_series.csv' и записать его в таблицу 'dft'.

In [209]:
dft = pd.read_csv('time_series.csv') # Читаем файл time_series.csv с помощью метода read_csv
dft.head() # Выводим первые 5 строк таблицы с помощью метода head.

Date   Time  direction_centre  direction_region
0  14/07/2011  17:44                21                46
1  14/07/2011  17:59                17                60
2  14/07/2011  18:14                13                48
3  14/07/2011  18:29                22                48
4  14/07/2011  18:44                15                51

На основе информаци о дате и времени создать в таблице dft столбец 'Date_Time'.

Элемент нового столбца должен иметь следующий вид:
14/07/2011	17:44

In [210]:
dft['Date_Time'] = dft['Date'] + ' ' + dft['Time'] # Создаем новый столбец Date_Time в таблице dft
dft.head() # Выводим первые 5 строк таблицы с помощью метода head.

Date   Time  direction_centre  direction_region         Date_Time
0  14/07/2011  17:44                21                46  14/07/2011 17:44
1  14/07/2011  17:59                17                60  14/07/2011 17:59
2  14/07/2011  18:14                13                48  14/07/2011 18:14
3  14/07/2011  18:29                22                48  14/07/2011 18:29
4  14/07/2011  18:44                15                51  14/07/2011 18:44

Удалить столбцы 'Date' и 'Time'.

In [211]:
dft = dft.drop(columns=['Date', 'Time']) # Удаляем столбцы Date и Time из таблицы dft с помощью метода drop
dft.head() # Выводим первые 5 строк таблицы с помощью метода head

direction_centre  direction_region         Date_Time
0                21                46  14/07/2011 17:44
1                17                60  14/07/2011 17:59
2                13                48  14/07/2011 18:14
3                22                48  14/07/2011 18:29
4                15                51  14/07/2011 18:44

Преобразовать тип переменной 'Date_Time' в формат `datetime`.


In [212]:
dft['Date_Time'] = pd.to_datetime(dft['Date_Time'], format='%d/%m/%Y %H:%M') # Преобразуем столбец Date_Time в тип datetime с помощью метода to_datetime
dft.head() # Выводим первые 5 строк таблицы с помощью метода head.

direction_centre  direction_region           Date_Time
0                21                46 2011-07-14 17:44:00
1                17                60 2011-07-14 17:59:00
2                13                48 2011-07-14 18:14:00
3                22                48 2011-07-14 18:29:00
4                15                51 2011-07-14 18:44:00

Установить столбец 'Date_Time' в качестве индекса.

In [213]:
dft = dft.set_index('Date_Time') # Устанавливаем столбец Date_Time в качестве индекса таблицы df с помощью метода set_index.
dft.head() # Выводим первые 5 строк таблицы с помощью метода head.

direction_centre  direction_region
Date_Time                                              
2011-07-14 17:44:00                21                46
2011-07-14 17:59:00                17                60
2011-07-14 18:14:00                13                48
2011-07-14 18:29:00                22                48
2011-07-14 18:44:00                15                51

Добавить признак 'both' с суммарным потоком велосипедистов.

In [214]:
dft = dft.assign(both=dft['direction_centre'] + dft['direction_region']) # Создаем новый столбец both в таблице dft, соединяя значения столбцов direction_centre и direction_region
dft.head() # Выводим первые 5 строк таблицы с помощью метода head

direction_centre  direction_region  both
Date_Time                                                    
2011-07-14 17:44:00                21                46    67
2011-07-14 17:59:00                17                60    77
2011-07-14 18:14:00                13                48    61
2011-07-14 18:29:00                22                48    70
2011-07-14 18:44:00                15                51    66

Добавить столбец 'weekday' с названиями дней недели на русском языке.

In [215]:
dft['weekday'] = dft.index.day_name() # Создаем новый столбец weekday в таблице dft, содержащий название дня недели для каждой строки.

english_to_russian = { # Создаем словарь english_to_russian, содержащий перевод названий дней недели с английского на русский язык
    'Monday': 'Понедельник',
    'Tuesday': 'Вторник',
    'Wednesday': 'Среда',
    'Thursday': 'Четверг',
    'Friday': 'Пятница',
    'Saturday': 'Суббота',
    'Sunday': 'Воскресенье'
}

dft['weekday'] = dft['weekday'].map(english_to_russian) # Применяем метод map к столбцу weekday, чтобы заменить названия дней недели на русский язык.
dft.head() # Выводим первые 5 строк таблицы с помощью метода head

direction_centre  direction_region  both  weekday
Date_Time                                                             
2011-07-14 17:44:00                21                46    67  Четверг
2011-07-14 17:59:00                17                60    77  Четверг
2011-07-14 18:14:00                13                48    61  Четверг
2011-07-14 18:29:00                22                48    70  Четверг
2011-07-14 18:44:00                15                51    66  Четверг

Выбрать временные интервалы, для которых суммарный поток менее 3 велосипедистов.

In [216]:
dft[dft['both'] < 3] # Выбираем строки из таблицы dft, в которых значение столбца both меньше 3

direction_centre  direction_region  both  weekday
Date_Time                                                             
2011-07-15 02:29:00                 0                 0     0  Пятница
2011-07-15 02:44:00                 0                 1     1  Пятница
2011-07-15 02:59:00                 1                 0     1  Пятница
2011-07-15 03:14:00                 0                 0     0  Пятница
2011-07-15 03:29:00                 1                 1     2  Пятница
...                               ...               ...   ...      ...
2017-01-17 04:39:00                 1                 1     2  Вторник
2017-01-17 04:54:00                 0                 1     1  Вторник
2017-01-17 05:24:00                 0                 1     1  Вторник
2017-01-18 03:24:00                 2                 0     2    Среда
2017-01-18 05:39:00                 1                 0     1    Среда

[8900 rows x 4 columns]

Выбрать временные интервалы, для которых в одном из направлений проехало менее 5 велосипедистов.

In [217]:
dft[(dft['direction_centre'] < 5) | (dft['direction_region'] < 5)] # Выбираем строки из таблицы dft, в которых значение столбца direction_centre меньше 5

direction_centre  direction_region  both  weekday
Date_Time                                                             
2011-07-14 21:29:00                 3                10    13  Четверг
2011-07-14 21:44:00                 7                 4    11  Четверг
2011-07-14 22:14:00                 4                 3     7  Четверг
2011-07-14 23:29:00                 3                 7    10  Четверг
2011-07-14 23:59:00                 1                13    14  Четверг
...                               ...               ...   ...      ...
2017-01-18 06:39:00                 2                 2     4    Среда
2017-01-18 06:54:00                 8                 2    10    Среда
2017-01-18 07:09:00                 9                 2    11    Среда
2017-01-18 07:24:00                11                 4    15    Среда
2017-01-18 07:39:00                24                 3    27    Среда

[53512 rows x 4 columns]

Определить среднее количество велосипедистов за 15 минут.

In [218]:
dft['total'] = dft['direction_centre'] + dft['direction_region'] # Создаем новый столбец total в таблице dft, содержащий сумму значений столбцов direction_centre и direction_region
dft['total'].mean() # Вычисляем среднее значение столбца total с помощью метода mean.

50.88382163209661

Определить среднее количество велосипедистов за 1 час.

In [219]:
dft['total'].resample('h').sum().mean()
# Группируем данные по часам с помощью метода resample.
# Суммируем значения столбца total в каждой группе с помощью метода sum.
# Вычисляем среднее значение суммы столбца total в каждой группе с помощью метода mean.

201.4091219219716

Найти даты с 5-ю самыми большими значениями количества велосипедистов, проехавших в направлении к центру.

In [220]:
s_df = dft.sort_values(by='direction_centre', ascending=False) # Сортируем строки таблицы df по убыванию значений столбца direction_centre с помощью метода sort_values.
s_df.head(5) # Выводим первые 5 строк отсортированной таблицы с помощью метода head.

direction_centre  direction_region  both  weekday  total
Date_Time                                                                    
2013-10-16 21:38:00               361                43   404    Среда    404
2014-09-23 08:12:00               333               124   457  Вторник    457
2016-10-06 09:16:00               325               155   480  Четверг    480
2011-10-04 08:27:00               319                88   407  Вторник    407
2013-10-03 08:49:00               318                77   395  Четверг    395

Найти дату с максимальным количеством велосипедистов за один день в обоих направлениях.

In [222]:
dft = dft.reset_index()
dft.groupby(dft['Date_Time'].dt.date)['total'].sum().idxmax()
# Сбрасываем индекс таблицы df с помощью метода reset_index.
# Группируем данные по датам с помощью метода groupby.
# Суммируем значения столбца total в каждой группе с помощью метода sum.
# Выбираем дату с максимальной суммой столбца total с помощью метода idxmax.

datetime.date(2016, 9, 29)